src/main/java/Objects/GameMapImpl.java

```diff
@@ -349,14 +353,18 @@ private synchronized void nextRound() {
         final boolean captureAwardMineral = awardMineral;
         mines.stream().filter((mine) -> (mine.getOwner() != null)).forEach((mine) -> {
             if(captureAwardMineral) {
-                final long amountMined = mine.mineResources(Configurations.getMineResourceAmount());
-                userMinerals.put(mine.getOwner().getName(), Math.min(userCapacity.get(mine.getOwner().getName()), amountMined + userMinerals.get(mine.getOwner().getName())));
+                final long amountMined = mine.mineResources(Configurations.getMineResourceAmount() - Configurations.getMineResourceReplenishAmount());
+
+                // if we used up all the resources, the mine dies
+                if (amountMined == 0l) {
+                    mine.setOwner(null);
+                    mines.remove(mine);

```

Comment:
I don't think you are allowed to remove the mine inside of the loop. You nullify the iterator no?

Reply:
It's okay because this is a copy on write  set.